File: netlimit/elevate.py
"""Privilege escalation helpers (sudo re-exec with absolute paths)."""

from __future__ import annotations

import os
import shutil
import sys
from pathlib import Path


def python_executable() -> str:
    """Absolute path to the interpreter that has netlimit installed."""
    return str(Path(sys.executable).resolve())


def console_script() -> Path | None:
    """Path to the netlimit console script, if we were launched via it."""
    script = Path(sys.argv[0]).resolve()
    if script.is_file() and os.access(script, os.X_OK) and script.name.startswith(
        "netlimit"
    ):
        return script
    candidate = Path(python_executable()).parent / "netlimit"
    if candidate.is_file() and os.access(candidate, os.X_OK):
        return candidate
    return None


def sudo_argv(*forwarded: str) -> list[str]:
    """Build a sudo argv that works when `netlimit` is not on secure_path."""
    script = console_script()
    if script is not None:
        return ["sudo", str(script), *forwarded]
    return ["sudo", python_executable(), "-m", "netlimit", *forwarded]


def sudo_command_hint(*forwarded: str) -> str:
    return " ".join(sudo_argv(*forwarded))


def elevate_or_exit(*forwarded: str) -> None:
    """Re-exec under sudo using absolute paths, or print a clear fallback and exit."""
    argv = sudo_argv(*forwarded)
    cmd = " ".join(argv)
    script = console_script()
    link_src = script or Path(python_executable()).parent / "netlimit"

    if shutil.which("sudo") is None:
        print("error: sudo not found. Re-run NetLimit as root:", file=sys.stderr)
        print(f"  {python_executable()} -m netlimit", file=sys.stderr)
        raise SystemExit(1)

    print("NetLimit needs root to change traffic rules.")
    print(f"Running: {cmd}")
    print()
    print("Tip: if sudo says “command not found”, install a system link once:")
    print(f"  sudo ln -sf {link_src} /usr/local/bin/netlimit")
    print()

    try:
        os.execvp("sudo", argv)
    except OSError as exc:
        print(f"error: failed to exec sudo: {exc}", file=sys.stderr)
        print(f"Run manually:\n  {cmd}", file=sys.stderr)
        raise SystemExit(1) from exc

File: src/elevate.rs
//! Re-exec under sudo with absolute paths (sudo secure_path safe).

use std::env;
use std::os::unix::process::CommandExt;
use std::path::PathBuf;
use std::process::Command;

use anyhow::{bail, Context, Result};

pub fn elevate(forwarded: &[String]) -> Result<()> {
    let sudo = which("sudo").context("sudo not found on PATH")?;
    let exe = env::current_exe().context("cannot resolve current executable")?;

    eprintln!("NetLimit needs root to change traffic rules.");
    eprintln!("Running: sudo {} {}", exe.display(), forwarded.join(" "));
    eprintln!();
    eprintln!("Tip: install a system link so `sudo netlimit` works:");
    eprintln!("  sudo ln -sf {} /usr/local/bin/netlimit", exe.display());
    eprintln!();

    let err = Command::new(sudo).arg(&exe).args(forwarded).exec();
    bail!("failed to exec sudo: {err}");
}

fn which(cmd: &str) -> Option<PathBuf> {
    env::var_os("PATH").and_then(|paths| {
        env::split_paths(&paths).find_map(|dir| {
            let p = dir.join(cmd);
            p.is_file().then_some(p)
        })
    })
}
